docs/rfcs/021-metering.md

```diff
@@ -0,0 +1,175 @@
+
+* While we don't yet have the autoscaling console can report `cpu time` as the number of seconds since the last `start_compute` event.
+
+* When we have autoscaling, `autoscaler-agent` can report `cpu time`*`compute_units_count` in the same increments as the proxy reports traffic. 
+
+Example event:
+
+```json
+{
+    "metric": "effective_compute_seconds",
+    "type": "increment",
+    "endpoint_id": "blazing-warrior-34",
+    "event_start_time": ...,
+    "event_stop_time": ...,
+    "value": 12345454,
+}
+```
+
+I'd also suggest reporting one value, `cpu time`*`compute_units_count`, instead of two separate fields as it makes event schema simpler (it is possible to treat it the same way as traffic) and preserves additivity.
+
+## Pageserver
+
+Pageserver knows / has access to / can calculate the rest of the metrics:
+
+* Written size -- that is basically `last_received_lsn`,
+* Synthetic storage size -- there is a way to calculate it, albeit a costly one,
+* Real storage size -- there is a way to calculate it using a layer map or filesystem,
+* S3 storage size -- can calculate it by S3 API calls
+
+Some of those metrics are expensive to calculate, so the reporting period here is driven mainly by implementation details. We can set it to, for example, once per hour. Not a big deal since the pageserver is stateful, and all metrics can be reported as an absolute value, not increments. At the same time, a smaller reporting period improves UX, so it would be good to have something more real-time.
+
+`written size` is primarily a safekeeper-related metric, but since it is available on both pageserver and safekeeper, we can avoid reporting anything from the safekeeper.
+
+Example event:
+
+```json
+{
+    "metric": "s3_tenant_size_bytes",
+    "type": "absolute",
+    "tenant_id": "5d07d9ce9237c4cd845ea7918c0afa7d",
+    "time": ...,
+    "value": 12345454,
+}
+```
+
+# Data collection
+
+## Push vs. pull
+
+We already have pull-based Prometheus metrics, so it is tempting to use them here too. However, in our setup, it is hard to tell when some metric changes. For example, garbage collection will constantly free some disk space over a week, even if the project is down for that week. We could also iterate through all existing tenants/branches/endpoints, but that means some amount of code to do that properly and most likely we will end up with some per-metric hacks in the collector to cut out some of the tenants that are surely not changing that metric.
+
+With the push model, it is easier to publish data only about actively changing metrics -- pageserver knows when it performs s3 offloads, garbage collection and starts/stops consuming data from the safekeeper; proxy knows about connected clients; console / autoscaler-agent knows about active cpu time.
+
+Hence, let's go with a push-based model.
```

Comment:
i suggest to design for both push and pull clients.

we can totally use push for our source of truth for billing, but support the pull readers too.
e.g. for expensive calculations, whenever they are triggered by an internal routine or by an external call, we could cache locally the latest value and the time when it was calculated. we can provided this into to pull-clients for debugging purposes or for ensuring the eventual consistency.

IOW, even though console or billing vendor would receive push, we would have the last measurement cached and available cheap and easy on demand for reporting to prom or to any other client via the API.

towards prometheus we could report something like last_known_{metric} and last_known_{metric}_age

Reply:
makes sense, let's keep Prometheus endpoints as well. That will also help to structure collection code on the source side -- that could be an usual metric which is periodically manually pushed somewhere.

Regarding the `last_known_*` I think that applies only to synthetic size. With all other metrics we can update them live.